libheif/CMakeLists.txt

```diff
@@ -1,84 +1,98 @@
-    heif_encoder_x265.h
-  )
-  set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${X265_CFLAGS}")
-  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${X265_CFLAGS}")
-
-  if (NOT "${X265_LIBRARY_DIRS}" STREQUAL "")
-    set(X265_LINKDIR "-L${X265_LIBRARY_DIRS}")
-  endif()
-endif()
+add_library(${PROJECT_NAME} SHARED
+    bitstream.cc
+    box.cc
+    error.cc
+    heif.cc
+    heif_context.cc
+    heif_file.cc
+    heif_image.cc
+    heif_hevc.cc
+    heif_colorconversion.cc
+    heif_plugin_registry.cc
+    heif_plugin.cc
+    ${libheif_headers}
+)
 
-# set(CMAKE_VERBOSE_MAKEFILE on)
+# Needed to find libheif/heif_version.h while compiling the library
+target_include_directories(${PROJECT_NAME} PRIVATE ${PROJECT_BINARY_DIR})
 
-add_definitions(-DHAVE_VISIBILITY)
-add_definitions(-DLIBHEIF_EXPORTS)
+target_compile_definitions(${PROJECT_NAME}
+    PUBLIC 
+        LIBHEIF_EXPORTS
+        HAVE_VISIBILITY
+)
 
-add_library(${LIBHEIF_LIBRARY_NAME} SHARED ${libheif_sources})
 if(LIBDE265_FOUND)
-  target_link_libraries(${LIBHEIF_LIBRARY_NAME} ${LIBDE265_LIBRARIES} ${LIBDE265_LINKDIR})
+    target_compile_definitions(${PROJECT_NAME} PRIVATE HAVE_LIBDE265=1)
+    target_sources(${PROJECT_NAME} PRIVATE
+        heif_decoder_libde265.cc
+        heif_decoder_libde265.h
+    )
+
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${LIBDE265_CFLAGS}")
+
+    if (NOT "${LIBDE265_LIBRARY_DIRS}" STREQUAL "")
+        set(LIBDE265_LINKDIR "-L${LIBDE265_LIBRARY_DIRS}")
+    endif()
+
+    target_link_libraries(${PROJECT_NAME} PRIVATE ${LIBDE265_LIBRARIES} ${LIBDE265_LINKDIR})
+
+    install(FILES heif_decoder_libde265.h DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/heif)
 endif()
+
 if(X265_FOUND)
-  target_link_libraries(${LIBHEIF_LIBRARY_NAME} ${X265_LIBRARIES} ${X265_LINKDIR})
-endif()
-if(CMAKE_COMPILER_IS_GNUCXX)
-  set_target_properties(${LIBHEIF_LIBRARY_NAME} PROPERTIES CXX_VISIBILITY_PRESET hidden)
+    target_compile_definitions(${PROJECT_NAME} PRIVATE HAVE_X265=1)
+    target_sources(${PROJECT_NAME} PRIVATE
+        heif_encoder_x265.cc
+        heif_encoder_x265.h
+    )
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${X265_CFLAGS}")
+
+    if (NOT "${X265_LIBRARY_DIRS}" STREQUAL "")
+        set(X265_LINKDIR "-L${X265_LIBRARY_DIRS}")
+    endif()
+
+    target_link_libraries(${PROJECT_NAME} PRIVATE ${X265_LIBRARIES} ${X265_LINKDIR})
+
+    install(FILES heif_encoder_x265.h DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/heif)
 endif()
 
-if(CMAKE_SYSTEM_PROCESSOR STREQUAL "x86_64")
-  SET_TARGET_PROPERTIES(${LIBHEIF_LIBRARY_NAME} PROPERTIES COMPILE_FLAGS "-fPIC")
-endif(CMAKE_SYSTEM_PROCESSOR STREQUAL "x86_64")
+write_basic_package_version_file(heifConfigVersion.cmake COMPATIBILITY ExactVersion)
+
+install(TARGETS ${PROJECT_NAME} EXPORT heifConfig
+    RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
+    LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+    ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
+)
+
+install(FILES ${libheif_headers} DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/heif)
```

Comment:
question: Should the header files be installed in a folder named `heif` or `libheif` ? I would not like to add inconsistencies between the CMake and the autotools/autoconfigure mode.

Reply:
The headers should be installed in a `libheif` directory.

Reply:
I changed this in the latest commits. I noticed while testing with the conan package that indeed I needed `libheif`